app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayAdapter.kt

```diff
@@ -0,0 +1,123 @@
+import com.rarepebble.colorpicker.ColorPickerView
+
+class UserDisplayAdapter(val entries: MutableList<UserDisplayItem>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
+
+    inner class EntryViewHolder(val binding: UserDisplayItemBinding) : RecyclerView.ViewHolder(binding.root) {
+        init {
+            binding.userDisplayDelete.setOnClickListener {
+                entries.removeAt(bindingAdapterPosition)
+                notifyItemRemoved(bindingAdapterPosition)
+            }
+
+            binding.userDisplayPickColorButton.setOnClickListener {
+                val item = binding.userDisplay ?: return@setOnClickListener
+                val picker = ColorPickerView(binding.root.context)
+                picker.showAlpha(false)
+
+                picker.color = item.color.toARGBInt() // to correctly set initial alpha to 255
+                MaterialAlertDialogBuilder(binding.root.context)
+                    .setView(picker)
+                    .setTitle(binding.root.context.getString(R.string.pick_custom_user_color_title))
+                    .setNegativeButton(binding.root.context.getString(R.string.dialog_cancel)) { _, _ -> }
+                    .setPositiveButton(binding.root.context.getString(R.string.dialog_ok)) { _, _ ->
+                        val pickedColor = picker.color
+                        item.color = pickedColor
+                        binding.userDisplayPickColorButton.setColorAndBg(item.color)
+                    }
+                    .show()
+            }
+
+            (binding.userDisplayEnableColor).apply {
+                setOnCheckedChangeListener { _, checked ->
+                    val item = binding.userDisplay ?: return@setOnCheckedChangeListener
+                    item.colorEnabled = checked
+                    binding.userDisplayPickColorButton.isVisible = checked
+                    binding.userDisplayPickColorButton.setColorAndBg(item.color)
+                }
+            }
+            (binding.userDisplayEnableAlias).apply {
+                setOnCheckedChangeListener { _, checked ->
+                    val item = binding.userDisplay ?: return@setOnCheckedChangeListener
+                    item.aliasEnabled = checked
+                    binding.userDisplayAliasInput.isVisible = checked
+                }
+            }
+        }
+    }
+
+    /** set text, text color, background color to represent specified color */
+    @SuppressLint("SetTextI18n")
+    private fun MaterialButton.setColorAndBg(colorRGB: Int) {
+        text = "#" + colorRGB.toHexCode()
+        setTextColor(colorRGB.toARGBInt().getContrastTextColor())
+        setBackgroundColor(colorRGB.toARGBInt())
+    }
+
+    inner class AddItemViewHolder(val binding: AddItemBinding) : RecyclerView.ViewHolder(binding.root) {
+        init {
+            binding.multiEntryAdd.setOnClickListener {
+                // ID 0, so that the create call generate the ID
```

Comment:
Too much logic for the UI, should be delegated to the viewmodel which then calls the repository.